src/Microsoft.Maintainability.Analyzers/Core/AvoidUninstantiatedInternalClasses.cs

```diff
@@ -37,6 +43,205 @@ public override void Initialize(AnalysisContext analysisContext)
+                    if (namedType != null)
+                    {
+                        instantiatedTypes.Add(namedType);
+                    }
+                }, OperationKind.ObjectCreationExpression);
+
+                startContext.RegisterSymbolAction(context =>
+                {
+                    INamedTypeSymbol type = (INamedTypeSymbol)context.Symbol;
+                    if (type.GetResultantVisibility() != SymbolVisibility.Public && !IsOkToBeUnused(type, context.Compilation))
+                    {
+                        internalTypes.Add(type);
+                    }
+                }, SymbolKind.NamedType);
+
+                startContext.RegisterCompilationEndAction(context =>
+                {
+                    IEnumerable<INamedTypeSymbol> uninstantiatedInternalTypes = internalTypes
+                        .Except(instantiatedTypes)
+                        .Where(type => !HasInstantiatedNestedType(type, instantiatedTypes));
+
+                    foreach (INamedTypeSymbol type in uninstantiatedInternalTypes)
+                    {
+                        context.ReportDiagnostic(type.CreateDiagnostic(Rule, type.FormatMemberName()));
+                    }
+                });
+            });
+        }
+
+        private bool AssemblyExposesInternals(Compilation compilation)
+        {
+            ISymbol assemblySymbol = compilation.Assembly;
+            ImmutableArray<AttributeData> attributes = assemblySymbol.GetAttributes();
+            return attributes.Any(
+                attr => attr.AttributeClass.Name.Equals("InternalsVisibleToAttribute", StringComparison.Ordinal));
+        }
+
+        private bool HasInstantiatedNestedType(INamedTypeSymbol type, List<INamedTypeSymbol> instantiatedTypes)
+        {
+            // We don't care whether a private nested type is instantiated, because if it
+            // is, it can only have happened within the type itself.
+            IEnumerable<INamedTypeSymbol> nestedTypes = type.GetTypeMembers()
+                .Where(member => member.DeclaredAccessibility != Accessibility.Private);
+
+            foreach (var nestedType in nestedTypes)
+            {
+                if (instantiatedTypes.Contains(nestedType))
+                {
+                    return true;
+                }
+
+                if (HasInstantiatedNestedType(nestedType, instantiatedTypes))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public static bool IsOkToBeUnused(INamedTypeSymbol type, Compilation compilation)
+        {
+            if (type.TypeKind != TypeKind.Class || type.IsAbstract)
+            {
+                return true;
+            }
+
+            // Attributes are not instantiated in IL but are created by reflection.
+            INamedTypeSymbol attributeSymbol = compilation.GetTypeByMetadataName("System.Attribute");
```

Comment:
Again, we should get this well known symbol in the compilation start action, and pass it as an argument to this method.